### ISSUE TITLE
Is it possible to update the log level?

### ISSUE THREAD

**Issue by nitro52** [structured-log/structured-log#33]
Hey, I'm looking at updating the structured-log-seq-sink to support dynamic log levels. When the sink logs a request the response will contain the current log level. You an also check at regular intervals allowing you to disable logging or increase the log level when debugging a problem in production and then scale it back to something like Error when done.

I'm thinking that the sink would call something that would update the min log level for the current logger.  I've been looking over the code and i don't think this is possible at the moment with the current architecture but i'm still trying to wrap my head around how it all fits together.  

So a few questions
1. Is it already possible for a sink to update the minLogLevel? 
2. Any suggestions on how you might support this? Perhaps something with the LevelMap?

**Comment by Wedvich** [comment]
There's no built-in support for updating the `minLogLevel` dynamically, but this is relatively straightforward to implement using a filter. Quick example:

```js
var dynamicLevelMap = new LevelMap('INFO');

var log = structuredLog.configure()
  .minLevel('VERBOSE')
  .filter(function (logEvent) {
    return dynamicLevelMap.isEnabled(logEvent.level);
  })
  .writeTo(consoleSink())
  .create();

log.debug('This should not get through');
dynamicLevelMap = new LevelMap('DEBUG');
log.debug('But this should');
```

`LevelMap` isn't exported so this won't work without modifications, but you can roll your own version of it for now (or duplicate the code from source). Dynamically updating the log level could be useful as a core part of the functionality though, so I think a PR would be welcome for this!

@ashleydavis thoughts?

**Comment by nitro52** [comment]
A filter sounds like a good implementation, though i think it would be nice if the sink could initialize this itself. The complexity of when and how to manage the log level shouldn't need to be maintained by the user. 

Is the LevelMap intended to be a single instance for each logger? It seems to be instantiated in createLogger  as though there would be, but it also gets instantiated minLevel. Why is that?

**Comment by nitro52** [comment]
@Wedvich 
I was looking at your ES6 implementation and noticed you beat me to adding the dynamicLevelMap. In this scenario the seq-sink would need a reference to the same levelMap as the one passed to minLevel. It seems a bit odd to have to write something like this though

```
   const levelMap = new LevelMap();
   const seqSink = new SeqSink(levelMap);

    const logger = new LoggerConfiguration()
      .minLevel(levelMap)
      .writeTo(logSink)
      .create();
```

because you need to pass the same instance into structured log and seqSink. I was thinking about updating the writeTo function to initialise the minLogLevel based on the sinks logLevel. Something like this. 

```
  writeTo(sinkOrLogger) {
    if (typeof sinkOrLogger === 'function') {
      sinkOrLogger = new WrappedSink(sinkOrLogger);
    }
    if(sinkOrLogger.levelMap instanceof LevelMap) {
      minLevel(sinkOrLogger.levelMap)
    }
    _pipelineStages.get(this).push(new SinkStage(sinkOrLogger));
    return this;
  }
```

what do you think?

**Comment by Wedvich** [comment]
This is now built after PR #35  in using the `DynamicLevelSwitch` class! :smile:

**Comment by Wedvich** [comment]
Closing this now that it's been released with [v0.1.0](https://github.com/structured-log/structured-log/releases/tag/v0.1.0).

**Comment by rburnham52** [comment]
@Wedvich nice work. I look forward to giving it a go when i get the time to jump back onto that project.
